test/functional/plugin/pack_spec.lua

```diff
@@ -1,40 +1,369 @@
+  git_add_commit('Initial commit for "defbranch"', 'defbranch')
+
+  -- Make `develop` the default remote branch
+  git_cmd({ 'checkout', '-b', 'develop' }, 'defbranch')
+
+  repo_write_file('defbranch', 'lua/defbranch.lua', 'return "defbranch develop"')
+  git_add_commit('Add to new default branch', 'defbranch')
+end
+
+local function setup_repo_gitsuffix()
+  init_test_repo('gitsuffix.git')
+
+  repo_write_file('gitsuffix.git', 'lua/gitsuffix.lua', 'return "gitsuffix main"')
+  git_add_commit('Initial commit for "gitsuffix"', 'gitsuffix.git')
+end
+
+local function setup_repo_semver()
+  init_test_repo('semver')
+
+  local add_tag = function(name)
+    repo_write_file('semver', 'lua/semver.lua', 'return "semver ' .. name .. '"')
+    git_add_commit('Add version ' .. name, 'semver')
+    git_cmd({ 'tag', name }, 'semver')
+  end
+
+  add_tag('v0.0.1')
+  add_tag('v0.0.2')
+  add_tag('v0.1.0')
+  add_tag('v0.1.1')
+  add_tag('v0.2.0-dev')
+  add_tag('v0.2.0')
+  add_tag('v0.3.0')
+  add_tag('0.3.1')
+  add_tag('v0.4')
+  add_tag('v0.2.1') -- Intentionally add version not in order
+  add_tag('v1.0.0')
+end
+
+-- Tests ======================================================================
+
 describe('vim.pack', function()
+  setup(function()
+    n.clear()
+    setup_repo_basic()
+    setup_repo_plugindirs()
+    setup_repo_pluginerr()
+    setup_repo_defbranch()
+    setup_repo_gitsuffix()
+    setup_repo_semver()
+  end)
+
+  before_each(function()
+    n.clear()
+  end)
+
+  after_each(function()
+    vim.fs.rm(pack_get_dir(), { force = true, recursive = true })
+  end)
+
+  teardown(function()
+    vim.fs.rm(repos_dir, { force = true, recursive = true })
+  end)
+
   describe('add()', function()
-    pending('works', function()
-      -- TODO
+    it('installs at proper version', function()
+      local out = exec_lua(function()
+        vim.pack.add({
+          { src = repos_src.basic, version = 'feat-branch' },
+        })
+        -- Should have plugin available immediately after
+        return require('basic')
+      end)
+
+      eq('basic feat-branch', out)
+
+      local rtp = vim.tbl_map(t.fix_slashes, n.api.nvim_list_runtime_paths())
+      local plug_path = pack_get_plug_path('basic')
+      local after_dir = vim.fs.joinpath(plug_path, 'after')
+      eq(true, vim.tbl_contains(rtp, plug_path))
+      -- No 'after/' directory in runtimepath because it is not present in plugin
+      eq(false, vim.tbl_contains(rtp, after_dir))
     end)
 
-    pending('reports errors after loading', function()
-      -- TODO
-      -- Should handle (not let it terminate the function) and report errors from pack_add()
+    it('can install from the Internet', function()
+      t.skip(fn.getenv('TEST_INTEG') == vim.NIL, 'Integration test: requires Internet connection')
+      exec_lua(function()
+        vim.pack.add({ 'https://github.com/neovim/nvim-lspconfig' })
+      end)
+      eq(true, exec_lua('return pcall(require, "lspconfig")'))
```

Comment:
Added a separate test for `vim.pack.add()` to be able to install from the Internet. As suggested [here](https://github.com/neovim/neovim/issues/34767#issuecomment-3036644145), this is done behind a special `TEST_INTEG` environment variable, which is enabled on CI. Seems to be working.